### PR TITLE
Error handling: when saving a datastack with relative paths fails

### LINKED ISSUE #1608: Handle saving datastack with relative paths, across drives/volumes
When creating a datastack JSON file, users can choose relative or absolute paths for files. Relative paths are relative to the save location of the JSON file. So when that is on a different drive from one of the files in the datastack, it raises,

```python
Traceback (most recent call last):
  File "flask\app.py", line 1473, in wsgi_app
  File "flask\app.py", line 882, in full_dispatch_request
  File "flask_cors\extension.py", line 165, in wrapped_function
  File "flask\app.py", line 880, in full_dispatch_request
  File "flask\app.py", line 865, in dispatch_request
  File "natcap\invest\ui_server.py", line 186, in write_parameter_set_file
  File "natcap\invest\datastack.py", line 568, in build_parameter_set
  File "natcap\invest\datastack.py", line 541, in _recurse
  File "natcap\invest\datastack.py", line 541, in <genexpr>
  File "natcap\invest\datastack.py", line 555, in _recurse
  File "<frozen ntpath>", line 791, in relpath
ValueError: path is on mount 'C:', start on mount 'D:'
```
We need to handle this more gracefully, at least when the Flask application is the caller, because we don't want to crash the app. I think it's okay for `datastack.py` to raise this exception.

We could refuse to create the datastack altogether. Or we could revert to using absolute paths.

### PR DESCRIPTION
**Pull request by emilyanndavis**
## Description
Fixes #1608 by logging and rendering an error message on failure to save a datastack with relative paths (e.g., when [os.path.relpath](https://docs.python.org/3/library/os.path.html#os.path.relpath) raises a ValueError on Windows).

relpath-failure_2024-09-17

## Checklist
- [x] Updated HISTORY.rst and link to any relevant issue (if these changes are user-facing)
~~- [ ] Updated the user's guide (if needed)~~
- [x] Tested the Workbench UI on macOS (by forcing `datastack.build_parameter_set` to raise a ValueError)
- [x] Tested the Workbench UI on Windows (thanks, @davemfish!)

### CODE REVIEW

**Review by davemfish — DISMISSED** [review]
@emilyanndavis This is working exactly as described for me on Windows!

It's great to have this new error-handling in place. I had a few minor comments about `try/except/else`, but in general everything looks very thorough good.

File: src/natcap/invest/datastack.py [hunk 1]
```diff
@@ -453,8 +457,12 @@ def build_datastack_archive(args, model_name, datastack_path):
...
     param_file_uri = os.path.join(temp_workspace,
                                   'parameters' + PARAMETER_SET_EXTENSION)
-    build_parameter_set(
-        rewritten_args, model_name, param_file_uri, relative=True)
+    try:
+        build_parameter_set(
+            rewritten_args, model_name, param_file_uri, relative=True)
+    except ValueError as message:
+        # Pass through for handling by ui_server
+        raise ValueError(message)
```
[the remark this answers is not in the document]
Comment (emilyanndavis) [reply src/natcap/invest/datastack.py:465]: Thank you! I knew this seemed off, but for some reason I felt compelled to explicitly pass the error through from one function to the next.

File: src/natcap/invest/datastack.py [hunk 2]
```diff
@@ -552,10 +563,19 @@ def _recurse(args_param):
...
+                    except ValueError:
+                        # On Windows, ValueError is raised when ``path`` and
+                        # ``start`` are on different drives
+                        raise ValueError(
+                            """Error: Cannot save datastack with relative
+                            paths across drives. Choose a different save
+                            location, or use absolute paths.""")
+                    else:
+                        # Always save unix paths.
+                        linux_style_path = temp_rel_path.replace('\\', '/')
```
[the remark this answers is not in the document]
Comment (emilyanndavis) [reply src/natcap/invest/datastack.py:578]: I can't remember what I read that made me think an `else` block was the way to go here. Your suggestion is simpler and seems to work just fine, thanks!

File: src/natcap/invest/ui_server.py [hunk 3]
```diff
@@ -221,15 +234,27 @@ def build_datastack_archive():
...
+            json.loads(payload['args']),
+            payload['moduleName'],
+            payload['filepath'])
+    except ValueError as message:
+        LOGGER.error(str(message))
+        return {
+            'message': str(message),
+            'error': True
+        }
+    else:
```
[the remark this answers is not in the document]
Comment (emilyanndavis) [reply src/natcap/invest/ui_server.py:253]: Updated these as well, thanks!

**Review by dcdenu4 — APPROVED** [review]
Alright, thanks @emilyanndavis! I did a quick lookover and Dave thought it was good to go as well.

**Review by dcdenu4 — APPROVED** [review]

**Pull request merged**